tiledb/sm/query/readers/aggregators/mean_aggregator.cc

```diff
@@ -0,0 +1,269 @@
+    throw MeanAggregatorStatusException("Result buffer doesn't exist.");
+  }
+
+  auto& result_buffer = buffers[output_field_name];
+  if (result_buffer.buffer_ == nullptr) {
+    throw MeanAggregatorStatusException(
+        "Mean aggregates must have a fixed size buffer.");
+  }
+
+  if (result_buffer.buffer_var_ != nullptr) {
+    throw MeanAggregatorStatusException(
+        "Mean aggregates must not have a var buffer.");
+  }
+
+  if (result_buffer.original_buffer_size_ != 8) {
+    throw MeanAggregatorStatusException(
+        "Mean aggregates fixed size buffer should be for one element only.");
+  }
+
+  bool exists_validity = result_buffer.validity_vector_.buffer();
+  if (field_info_.is_nullable_) {
+    if (!exists_validity) {
+      throw MeanAggregatorStatusException(
+          "Mean aggregates for nullable attributes must have a validity "
+          "buffer.");
+    }
+
+    if (*result_buffer.validity_vector_.buffer_size() != 1) {
+      throw MeanAggregatorStatusException(
+          "Mean aggregates validity vector should be for one element only.");
+    }
+  } else {
+    if (exists_validity) {
+      throw MeanAggregatorStatusException(
+          "Mean aggregates for non nullable attributes must not have a "
+          "validity "
+          "buffer.");
+    }
+  }
+}
+
+template <typename T>
+void MeanAggregator<T>::aggregate_data(AggregateBuffer& input_data) {
+  tuple<double, uint64_t, optional<uint8_t>> res{0, 0, nullopt};
+
+  bool overflow = false;
+  try {
+    if (input_data.is_count_bitmap()) {
+      res = mean<uint64_t>(input_data);
+    } else {
+      res = mean<uint8_t>(input_data);
+    }
+  } catch (std::overflow_error&) {
+    overflow = true;
+  }
+
+  {
+    // This might be called on multiple threads, the final result stored in sum_
+    // should be computed in a thread safe manner. The mutex also protects
+    // sum_overflowed_ which indicates when the sum has overflowed.
+    std::unique_lock lock(mean_mtx_);
+
+    // A previous operation already overflowed the sum, return.
+    if (sum_overflowed_) {
+      return;
+    }
+
+    // If we have an overflow, signal it, else it's business as usual.
+    if (overflow) {
+      sum_overflowed_ = true;
+      sum_ = std::get<0>(res);
+      count_ = std::numeric_limits<uint64_t>::max();
+      return;
+    } else {
+      // This sum might overflow as well.
+      try {
+        safe_sum(std::get<0>(res), sum_);
+        safe_sum(std::get<1>(res), count_);
+      } catch (std::overflow_error&) {
+        sum_overflowed_ = true;
+      }
+    }
+  }
+
+  if (field_info_.is_nullable_ && std::get<2>(res).value() == 1) {
+    validity_value_ = 1;
+  }
+}
+
+template <typename T>
+void MeanAggregator<T>::copy_to_user_buffer(
+    std::string output_field_name,
+    std::unordered_map<std::string, QueryBuffer>& buffers) {
+  auto& result_buffer = buffers[output_field_name];
+  *static_cast<double*>(result_buffer.buffer_) = sum_ / count_;
```

Comment:
This implementation needs to be changed to a standalone function which takes a container and returns a value, so that it can be unit-tested in isolation and *[not as a side effect of running a query](https://github.com/TileDB-Inc/TileDB/pull/4292/files#diff-a82a9e9c8ebb066fc765e5ed1bbe6723d8f901a6c2e9cafe467349190a243576R1424)*.

Reply:
I'm not sure I see the value here. This just stores the value in a buffer... It's easily unit testable.

Reply:
More, the reason a separate mean function was implemented is because we can process sum and count at the same time, which might be faster.

Reply:
> It's easily unit testable.
I disagree.

"Unit testable" here means a unit test runner inside `tiledb/query` somewhere; let's call it `unit_query_aggregate` for the moment. That test runner should not need to link the whole library, because it's isolated enough to link separately.  In particular, there should be no need to instantiate an object of `class Query` or of any reader class to test the operation.

This is a baseline of what a unit test would be: to have a unit test runner that minimizes additional code.

The existing test in `test-cppapi-aggregates.cc` is not a unit test. It's an integration test. It tests a fully integrated system across all layers of abstraction simultaneously. We want tests like this. They are not unit tests and they cannot replace unit tests.

Reply:
@eric-hughes-tiledb We do have unit_mean.cc that doesn't link the whole library. It does use QueryBuffer, but that's just a header getting pulled in. I can see the point for changing this in the future as a follow up work item, but this PR is merely using a mechanism already in place and I don't think it's the right time to go and change all the aggregators.

Reply:
>I don't think it's the right time to go and change all the aggregators

There won't ever be a right time. There's a large amount of copypasta in the current way these operators are being implemented. That will only get worse as we add more operators. The first piece of `dev` on this topic isn't even a week and a half old yet.

We need an implementation with better internal abstraction and greater code reuse. To pick up on one that was immediately apparent, we had essentially four duplicate versions of `validate_output_buffer` and this PR adds a fifth. That's a symptom of a design that needs reworking. Another symptom is that each operator has separate, explicit logic for handling nullable/nonnullable, bitmaps, etc. All that code should be generic.

To Isaiah's point, the existing tests do run without reference to a running query. His larger point is that we need to better decomposition in the implementation and we need independent tests for the components.

An operator should be able to be defined by only a small set things:
* Input arguments: number and type
* Output results: currently only 1 and dependent on
 a small set of functions, one each for eac

Reply:
>I don't think it's the right time to go and change all the aggregators

There won't ever be a right time. There's a large amount of copypasta in the current way these operators are being implemented. That will only get worse as we add more operators. The first piece of `dev` on this topic isn't even a week and a half old yet.

We need an implementation with better internal abstraction and greater code reuse. To pick up on one that was immediately apparent, we had essentially four duplicate versions of `validate_output_buffer` and this PR adds a fifth. That's a symptom of a design that needs overhaul. Another symptom is that each operator has separate, explicit logic for handling nullable/nonnullable, bitmaps, etc. All that code should be generic; it should be written once and that instantiated through template arguments.

To Isaiah's point, the existing tests do run without reference to a running query. His larger point is that we need to better decomposition in the implementation and we need independent tests for the components. To pick a particular example: we need to be able to test the arithmetic operations without assigning them to a field.

An operator should be able to be defined by only a small set things:
* Input arguments: number (currently either 0 or 1) and type
  * This includes which types are acceptable (and later which combinations of types)
* Output results: number (currently always 1) and type
* A small set of functions (and function templates) implementing the calculations.
Everything else in the code should be common for all aggregates.

Reply:
@eric-hughes-tiledb  validate_output_buffer is vastly different in the min max aggregator versus the sum aggregator... And it might be different for others too. So I wouldn't say there are 5 of the same implementation. I will agree it is duplicate here in this particular addition and the mean/sum aggregators should probably share a base class. I'll look into that. Where else in the aggregator code base do you see a lot of copy pasta?

I've already split a separate computation object (AggregateSum). I can make some more tests for it.

Reply:
> validate_output_buffer is vastly different 

They're not vastly different; they're all small variations on the same theme. The validity of an output buffer derives from a specification of an output type. Given an output type, the validation function follows. That validation function should be derived from a template somehow, either directly or from a class argument if validation is a member function.

Reply:
>Where else in the aggregator code base do you see a lot of copy pasta?

As I said above:
>explicit logic for handling nullable/nonnullable, bitmaps, etc.

There should be one place with this dispatch happens. It should use template arguments and all operators should use it in common.

Reply:
>separate computation object (AggregateSum)

It's a start. Turn that into a concept. Other computation objects should be able to satisfy the same concept. Building an operator shouldn't be much more that writing a computation class and then adding annotation data about its inputs and outputs.

Reply:
There's no need to rework toward this goal all in one piece. 
* Make function templates for particular skeletons of code. For each of them define a policy class to capture all the generic functionality.
* As the issues become clearer, combine the policy classes where they overlap.
* Derive classes from a class template instantiated with a policy class.
* Make a common base class (non-template) for virtual function definitions.

I'm guessing the class hierarchy will start to look like this:
```
class Operator {
   virtual size_t arity();
   // ...
};

template<class Policy>
class OperatorImpl : public Operator {
  // ... policy-dependent operations
};

class OperatorSum : public OperatorImpl<SumPolicy> {
  // ...
};
using OperatorCount  = OperatorImpl<CountPolicy>;
```
That last line, using an alias, is something like the desired final state, where all the variation has been captured through policy specification. `OperatorSum` is what a transitional class looks like during the rewrite.